File: Assets/src/model/enemies/Wildekin.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Chases you.\nStays one Tile away from Walls or non-Wildekins, but will attack you if possible.\nRuns away for three turns after it attacks.", flavorText: "")]
public class Wildekin : AIActor, IAttackHandler {
  public override float turnPriority => 60;
  public Wildekin(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 7;
  }

  internal override (int, int) BaseAttackDamage() => (3, 3);

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        // move one step closer
        var tiles = AdjacentTilesInPreferenceOrder();
        if (tiles.Any()) {
          var bestScore = TilePreference(tiles.First());
          var tile = tiles.Where(t => TilePreference(t) == bestScore)
            .OrderBy((t) => t.DistanceTo(GameModel.main.player))
            .First();
          return new MoveToTargetTask(this, tile.pos);
        } else {
          return new WaitTask(this, 1);
        }
      }
    } else {
      var tiles = AdjacentTilesInPreferenceOrder();
      if (tiles.Any()) {
        // tile preference version of moving randomly
        var bestScore = TilePreference(tiles.First());
        var tile = Util.RandomPick(tiles.Where(t => TilePreference(t) == bestScore));
        return new MoveToTargetTask(this, tile.pos);
      } else {
        return new WaitTask(this, 1);
      }
    }
  }

  public IEnumerable<Tile> AdjacentTilesInPreferenceOrder() => floor
    .GetAdjacentTiles(pos)
    .Where(t => (t.CanBeOccupied() || t == this.tile))
    .OrderBy(TilePreference);

  // lower = more preferred
  public int TilePreference(Tile t) => floor
    .GetAdjacentTiles(t.pos)
    .Where(t2 =>
      // Avoid walls
      t2 is Wall ||
      // Avoid non Wildekins
      (t2.body != null && !(t2.body is Wildekin))
    ).Count()
    // never walk directly into the Player
    + (t.IsNextTo(GameModel.main.player) ? 100 : 0);

  public void OnAttack(int damage, Body target) {
    if (target is Actor a) {
      SetTasks(new RunAwayTask(this, target.pos, 3, false));
    }
  }
}

[System.Serializable]
[ObjectInfo(description: "Applies Vulnerable when it hits you.\nGets stunned when attacked.", flavorText: "")]
public class Dizapper : AIActor, IAttackHandler, IBodyTakeAttackDamageHandler {
  public override float turnPriority => 60;
  public Dizapper(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 9;
  }

  internal override (int, int) BaseAttackDamage() => (1, 1);

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        return new ChaseTargetTask(this, player);
      }
    } else {
      return new MoveRandomlyTask(this);
    }
  }

  public void OnAttack(int damage, Body target) {
    if (target is Actor a) {
      a.statuses.Add(new VulnerableStatus(10));
    }
  }

  public void HandleTakeAttackDamage(int damage, int hp, Actor source) {
    if (damage > 0) {
      SetTasks(new WaitTask(this, 1));
      statuses.Add(new SurprisedStatus());
    }
  }
}

[System.Serializable]
[ObjectInfo(description: "When attacked, it duplicates into two with half HP.", flavorText: "")]
public class Goo : AIActor, IBodyTakeAttackDamageHandler {
  public override float turnPriority => 50;
  public Goo(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 20;
  }

  internal override (int, int) BaseAttackDamage() => (1, 1);

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        return new ChaseTargetTask(this, player);
      }
    } else {
      return new MoveRandomlyTask(this);
    }
  }

  public void HandleTakeAttackDamage(int damage, int hp, Actor source) {
    if (damage > 0) {
      SetTasks(new TelegraphedTask(this, 1, new GenericBaseAction(this, Split)));
    }
  }

  void Split() {
    var hp1 = Mathf.CeilToInt(hp / 2f);
    var hp2 = Mathf.FloorToInt(hp / 2f);
    var goo1 = new Goo(pos);
    goo1.hp = hp1;
    goo1.ClearTasks();
    var goo2 = new Goo(pos);
    goo2.hp = hp2;
    goo2.ClearTasks();
    floor.PutAll(goo1, goo2);
    KillSelf();
  }
}

[System.Serializable]
[ObjectInfo(description: "If the Hardshell would take 3 or more attack damage, that is instead reduced to 0.", flavorText: "")]
public class HardShell : AIActor, IAttackDamageTakenModifier {
  public override float turnPriority => 50;
  public HardShell(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 9;
  }

  internal override (int, int) BaseAttackDamage() => (2, 2);

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        return new ChaseTargetTask(this, player);
      }
    } else {
      return new MoveRandomlyTask(this);
    }
  }

  public int Modify(int input) {
    if (input >= 3) {
      input = 0;
    }
    return input;
  }
}

File: web/src/model/enemies/Wildekin.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackTask } from '../tasks/AttackTask';
import { MoveToTargetTask } from '../tasks/MoveToTargetTask';
import { WaitTask } from '../tasks/WaitTask';
import { RunAwayTask } from '../tasks/RunAwayTask';
import { ATTACK_HANDLER, type IAttackHandler } from '../Actor';
import { Wall, type Tile } from '../Tile';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { MyRandom } from '../../core/MyRandom';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Body } from '../Body';

/**
 * Chases you. Stays one tile away from walls or non-Wildekins.
 * Attacks you if possible, then runs away for 3 turns.
 * Port of C# Wildekin.cs.
 */
export class Wildekin extends AIActor implements IAttackHandler {
  readonly [ATTACK_HANDLER] = true as const;

  get turnPriority(): number { return 60; }

  constructor(pos: Vector2Int) {
    super(pos);
    this._hp = 7;
    this._baseMaxHp = 7;
  }

  baseAttackDamage(): [number, number] {
    return [3, 3];
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;
    if (this.canTargetPlayer()) {
      if (this.isNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        const tiles = this.adjacentTilesInPreferenceOrder();
        if (tiles.length > 0) {
          const bestScore = this.tilePreference(tiles[0]);
          const bestTiles = tiles.filter(t => this.tilePreference(t) === bestScore);
          const tile = bestTiles.sort(
            (a, b) => a.distanceTo(GameModelRef.main.player.pos) - b.distanceTo(GameModelRef.main.player.pos)
          )[0];
          return new MoveToTargetTask(this, tile.pos);
        } else {
          return new WaitTask(this, 1);
        }
      }
    } else {
      const tiles = this.adjacentTilesInPreferenceOrder();
      if (tiles.length > 0) {
        const bestScore = this.tilePreference(tiles[0]);
        const bestTiles = tiles.filter(t => this.tilePreference(t) === bestScore);
        const tile = MyRandom.Pick(bestTiles);
        return new MoveToTargetTask(this, tile.pos);
      } else {
        return new WaitTask(this, 1);
      }
    }
  }

  private adjacentTilesInPreferenceOrder(): Tile[] {
    return this.floor!
      .getAdjacentTiles(this.pos)
      .filter(t => t.canBeOccupied() || t === this.tile)
      .sort((a, b) => this.tilePreference(a) - this.tilePreference(b));
  }

  /** Lower = more preferred. */
  private tilePreference(t: Tile): number {
    const adjacentBad = this.floor!
      .getAdjacentTiles(t.pos)
      .filter(t2 =>
        t2 instanceof Wall ||
        (t2.body != null && !(t2.body instanceof Wildekin))
      ).length;
    const nearPlayer = t.isNextTo(GameModelRef.main.player) ? 100 : 0;
    return adjacentBad + nearPlayer;
  }

  onAttack(_damage: number, target: Body): void {
    if ('faction' in target) {
      this.setTasks(new RunAwayTask(this, target.pos, 3, false));
    }
  }
}

entityRegistry.register('Wildekin', Wildekin);
